geomfn: fix check of st_segmentize when number of segments overflow

otan opened the pull request:
Refs: #62919

Release note (sql change): Fix ST_Segmentize panicking when the number
of segments to generate exceeds math.MaxInt64.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63136)
<!-- Reviewable:end -->

rafiss commented on pkg/geo/geomfn/segmentize.go:
```diff
@@ -64,10 +64,19 @@ func segmentizeCoords(a geom.Coord, b geom.Coord, maxSegmentLength float64) ([]f
+	// With doubles, we can exceed the max int, so check that first.
+	if doubleNumPoints > float64(math.MaxInt64) {
+		return nil, errors.Newf(
+			"attempting to segmentize into too many coordinates",
```
can we have more info in the message like how line 82 has?

andyyang890 replied:
Maybe instead we could just move the check on line 80 up and only check it once? Presumably `float64(math.MaxInt64)` will always be greater than `geo.MaxAllowedSplitPoints`.

otan replied:
i avoided this because floats are ugly to format, but ok

rafiss reviewed (commented):
lgtm overall! thanks for the fix

andyyang890 reviewed (commented):
Do we need to fix the geogfn version too?

otan commented:
> Do we need to fix the geogfn version too?

no, because we can never create more than 34359738370 points (i think this is more or less the max resolution of lat/lngs. i couldn't reproduce the same error)

otan committed:
geomfn: fix check of st_segmentize when number of segments overflow

Release note (sql change): Fix ST_Segmentize panicking when the number
of segments to generate exceeds math.MaxInt64.

andyyang890 reviewed (approved):
LGTM!

otan commented:
bors r=andyyang890

rafiss commented:
@otan looks like the same thing can happen for GEOGRAPHY

```
SELECT st_segmentize('0105000060E61000000900000001020000400600000062F5D389A8255EC0E2F50BB9AB0E53C0C0F7367CD430FB41F03292D0C41545C03C5C5FEEC00C36C02C795F1FA572E04190B882C6B1C95E40D3070BB9C03342C054CCB5C7916BECC1BCD12C445D925F404DA0F761433640C0384C02FEC9E3E841D81909F2A15E4440908A40BFE92015C07A5D8879E842F6C18A5E456A01C96240D4D4DD67F9984E4044DFF8A3A470F7C10102000040040000004C9013CE422940C0080DE1B86C1B31C036EEAFD95EAFF8C14CE9A4F0B1A04040605B5893151408C0E4BB0E7D00B1E24104599BE8D0B95C407091AF36D09514404A820B37BA3BE0C10069A840096DF4BF1CB924E8D5053040386E254FFA27F941010200004004000000BDEB3DD3B1F550C0D2F16B6A5B3152C0948CC43CB330F2C1ECBA1A72586C5840DCC1285AA42C4BC01A76D9160B33FAC1984D59504B7660408E02EAED4FB33FC008989C951D12EF41A4DEE76266394440C08E91C16ADC43400D534F5319BA01C2010200004006000000D8CE8E61DC6E44C0F3DACCD020A343C0368D6BECDE7CFA418C1C511AE53F444050137FF45A3F1BC0208F617DEF0102424A50988F948F6540B0B8193019501CC00C0FBB700C3200C2A094BE6A6C896440ACACF544D54B4940C673DE174232004274580714D05355402CB21A0394B854401878F3AA2DA5D9C148211A5CA77C5DC01090575ED62253408E7737405ACC0042010200004007000000F3561800CDEF54C0DD986C7F9EEA55C077F8B656954EFFC1A84257DAA8B149C0FAA54F1D2EC450C06027734E181BCA413C9E37A3CB8240C046E3E6A89FCB55C0572A342C24D7FBC1F8B8C8C19A8A36C0BE0AE4212EE255C0B395405E7DA7F1C154EC6EB3449542401F0D08CBAC0F52C038BB7C7EBE37D7C1EA362D3B6C586340D6E11DD799194DC0C4F35D443DE7FEC1D2F6217DD4E4634080B83B66E16E35C0909B0F4E6F4401C2010200004005000000F814453F7F6F48C0D21100F0AAD652C0BAA0F5A69709F4C1603CEDAC9FE03A404803BFF4F8CD4DC0BBB4EDD035B0F5C1C8114851E14A5A403EC14F3C59A455C050E243898C36C0C16EBF2A5F0C346540B2074C41346047C0B47F72290D8EDEC110E75472467161404CBEF44BE851564041E7954E67E9F4C101020000400300000068D4BC8BCAD360408EDAF26314225040406C8E7AABCCB94114680A9037DB464010BD37A541AB49400D0EEF681B4FF2C100B77FD6DA7055C024078C93B69151405EE2A6F5AE4AFA41010200004006000000AD81F7020C4E59C00468A1A520AA51C008F9BD541014DC415E999D43557C624005196DA0D5C549C0F7B9CB02B39302C278FF560CFBE457406089A90FF5CF2F4090083D1D0CADF8415002C5B62A7C5C4038235652F9024A40A65D419B6BB9FCC1E0967A0F44B133402C167F6DD8583F405497314A9473F1415847CD0F91045340606D0436AFEB5240902EAB25E286C4C1010200004007000000C982C3CB9C7B60C05FCF369607F355C0601ACBBFA8C4E741008555378B4216C0B6628EAF2A334CC0D8F8C69CA567DD413C4AE0A671E75640F056D7AFF83B3DC0207A84A734F1DE41B423170550D1584076B0DF652CDC41409C060409B7F0F4C164280F5EB0DF50C09039E035D7B448406CC31C918C8CEF4179B091B84D8852C04C52F44B04DF3740C09DFB279E95BD416E5DCA897AF261C080EB55E9C385EA3F0465624B0C13EE41':::GEOGRAPHY, 1.401298464324817e-45:::FLOAT8)

ERROR: internal error: runtime error: makeslice: cap out of range
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/util/errorutil/catch.go:29: ShouldCatch()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:170: func1()
runtime/panic.go:975: gopanic()
runtime/slice.go:31: panicmakeslicecap()
runtime/slice.go:95: makeslice()
github.com/cockroachdb/cockroach/pkg/geo/geogfn/segmentize.go:98: segmentizeCoords()
github.com/cockroachdb/cockroach/pkg/geo/geosegmentize/geosegmentize.go:43: Segmentize()
github.com/cockroachdb/cockroach/pkg/geo/geosegmentize/geosegmentize.go:58: Segmentize()
github.com/cockroachdb/cockroach/pkg/geo/geogfn/segmentize.go:49: Segmentize()
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/geo_builtins.go:4948: func470()
```
